### DISCUSSION TITLE
Anti-aliasing of mask-based strokes

### DISCUSSION THREAD

**Discussion by hrdl-github** [xournalpp/xournalpp#4072] [General] [answered]
I would like to disable anti-aliasing for strokes that are being drawn interactively. Am I right that the anti-aliasing happens when drawing to the mask's surface? I've tried modifying https://github.com/xournalpp/xournalpp/blob/962fd4be4f40061462a75404447c5c469e56153e/src/core/control/tools/StrokeHandler.cpp#L370-L375
by adding a `cairo_set_antialias(cr, CAIRO_ANTIALIAS_NONE);`  statement as well as using `CAIRO_FORMAT_A1` instead of `CAIRO_FORMAT_A8`. What am I missing? Does the anti-aliasing happen (additionally?) at a later time?

Background: I'm using Xournal++ with an e-ink screen, which can react faster when from driving pixels either to white or to black (instead of grey).

**Comment by bhennion** [comment]
Is it reacting faster when writing a dashed stroke? If not, then the issue is not in this mask (which is not created for dashed or filled strokes).
You can try to set the page masks to CAIRO_FORMAT_A1, but I'm not sure this'll be enough. Those masks (one for each page of the document) are created in control/jobs/RenderJob.cpp. You can also try and turn antialiasing off when those masks are applied, in gui/PageView.cpp:paint.

**Reply by hrdl-github** [reply]
It's not that Xournal++ or cairo's anti-aliasing is slow -- when rendering only black and white contact the display controller can be instructed to use a faster waveform, which makes sense when rendering pen input. The other waveforms that allow more grey values introduce a noticeable lag. At the moment this mode still leads to artifacts, as all the grey pixels are not being rendered, which is why I want to disable this for ongoing strokes. I've managed to disable anti-aliasing for the finished strokes, but not the current one.

**Reply by bhennion** [reply]
Ah, maybe when the mask is blitted then, in StrokeHandler::draw. Maybe adding some `cairo_set_antialias(cr, CAIRO_ANTIALIAS_NONE);` around those lines
https://github.com/xournalpp/xournalpp/blob/962fd4be4f40061462a75404447c5c469e56153e/src/core/control/tools/StrokeHandler.cpp#L47-L50

**Comment by hrdl-github** [comment, marked as answer]
Adding a `cairo_set_antialias(cr, CAIRO_ANTIALIAS_NONE);` to e.g. https://github.com/xournalpp/xournalpp/blob/962fd4be4f40061462a75404447c5c469e56153e/src/core/control/tools/StrokeHandler.cpp#L370-L375 removes the anti-aliasing.

Outdated (before https://github.com/xournalpp/xournalpp/commit/8f8102c493f3fda8ee8a7f923a620592bdf618b1):
This still leaves some grey values, which I got rid of by replacing the call to `cairo_mask_surface` with
```
cairo_pattern_t *pattern = cairo_pattern_create_for_surface(mask->surf);
cairo_pattern_set_filter(pattern, CAIRO_FILTER_NEAREST);
cairo_mask(cr, pattern);
cairo_pattern_destroy(pattern);
```
This suggests that the mask's surface needs to be scaled, which I will try to avoid in the next iteration.

**Reply by bhennion** [reply]
> This suggests that the mask's surface needs to be scaled, which I will try to avoid in the next iteration.

It really should not. Please tell if and when you understand why it is recaled.

**Reply by hrdl-github** [reply]
You are right that it's not the scaling that makes the difference. Instead it's the fractional translation specified here: https://github.com/xournalpp/xournalpp/blob/962fd4be4f40061462a75404447c5c469e56153e/src/core/control/tools/StrokeHandler.cpp#L394-L395
I will simply round these instead in addition to disabling anti-aliasing for the mask's context.

**Reply by bhennion** [reply]
> I will simply round these instead in addition to disabling anti-aliasing for the mask's context.

I think it would be worth patching master with this change too.

**Reply by hrdl-github** [reply]
I will briefly test this without my modifications later and then make a pull requests.
